src/components/use-auth-state.ts

```diff
@@ -0,0 +1,356 @@
+import { AuthenticationService } from '../services/authentication-service'
+import { GlobalConfig } from '../global'
+import { Reducer, useReducer, useEffect, useMemo } from 'react'
+import * as Sentry from '@sentry/browser'
+
+interface AuthReducerState {
+  showSignIn: boolean
+  loading: boolean
+  authenticated: boolean
+  companyCode?: string
+  defaultCompanyCode?: string
+  companyCodeIsValid: boolean
+  authenticationService?: AuthenticationService
+  apiHost?: string
+}
+
+type AuthReducerAction =
+  | {
+      type:
+        | 'company-code-fetching'
+        | 'company-code-fetched-error'
+        | 'global-config-fetching'
+        | 'global-config-fetched-error'
+        | 'authentication-started'
+        | 'authentication-success'
+        | 'authentication-user-cancelled'
+        | 'authentication-failure'
+        | 'login-required'
+        | 'authentication-signout-started'
+        | 'authentication-signout-success'
+        | 'is-authenticated-starting'
+    }
+  | { type: 'company-code-fetched' | 'company-code-changed'; data: { companyCode?: string } }
+  | { type: 'global-config-fetched'; data: { globalConfig: GlobalConfig } }
+  | { type: 'is-authenticated-finished'; data: { isAuthenticated: boolean } }
+
+export type AuthenticationServiceFactory = (globalConfig: GlobalConfig) => AuthenticationService
+
+type AuthStateReducer = Reducer<AuthReducerState, AuthReducerAction>
+
+const createAuthStateReducerFn = (authenticationServiceFactory: AuthenticationServiceFactory): AuthStateReducer => {
+  return (state: AuthReducerState, action: AuthReducerAction): AuthReducerState => {
+    const message = `use-auth-state reducer action: ${action.type}`
+
+    if ('data' in action) {
+      Sentry.addBreadcrumb({ message: message, data: action.data })
+    } else {
+      Sentry.addBreadcrumb({ message: message })
+    }
```

Comment:
Is this absolutely terrible? :(
```suggestion
    Sentry.addBreadcrumb({ message: message, data: 'data' in action ? action.data : undefined })
```

Reply:
No I don't think so. I'll make that adjustment.